File: Java/Leetcode/BackTracking/WordSearch.java
package com.brianprojects.Leetcode.BackTracking;

//Time Complexity is: O(m*n)*(4^t) with t being the length of the string (target) | Space Complexity is: O(m*n)
public class WordSearch {
    public boolean exist(char[][] board, String word){
        int r = board.length;
        int c = board[0].length;

        for(int row = 0; row<board.length; row++){
            for(int col = 0; col<board[row].length; col++){
                if(word.charAt(0) == board[row][col] && wordExits(board, new int[r][c],word ,row, col, 0)){
                    return true;
                }
            }
        }
        return false;
    }

    private boolean wordExits(char[][] board, int[][] emptyBoard, String word,int row, int col, int index) {
        if(index >= word.length()){
            return true;
        }
        if(row <0 || row>board.length-1 || col<0 || col>board[row].length-1){
            return false;
        }
        if(board[row][col] != word.charAt(index) || emptyBoard[row][col] == 1){
            return false;
        }
        emptyBoard[row][col]++;
        if(!(wordExits(board, emptyBoard,word ,row-1, col, index+1)
                || wordExits(board, emptyBoard,word ,row+1, col, index+1)
                || wordExits(board, emptyBoard,word ,row, col-1, index+1)
                ||wordExits(board, emptyBoard,word ,row, col+1, index+1))){

            emptyBoard[row][col]--;
            return false;
        }
        return true;
    }

}



//////////////////////////////////////////////////////////////////////////////////////////////////
/*

private boolean wordExits(char[][] board, int[][] emptyBoard, String word,int row, int col, int index) {
        if(index >= word.length()){
            return true;
        }
        if(row <0 || row>board.length-1 || col<0 || col>board[row].length-1){
            return false;
        }
        for(int r = row; r<board.length; r++){
            for(int c = col; c<board[row].length; c++){
                if(word.charAt(index) == board[r][c]){
                    emptyBoard[row][col]++;
                    wordExits(board, emptyBoard,word, r-1, c, index+1);
                    wordExits(board, emptyBoard,word ,r+1, c, index+1);
                    wordExits(board, emptyBoard,word ,r, c-1, index+1);
                    wordExits(board, emptyBoard,word ,r, c+1, index+1);
                    emptyBoard[r][c]--;
                }
            }
        }
        return false;
    }
 */
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/*
if(index >= word.length()){
            return true;
        }
        if(row <0 || row>board.length-1 || col<0 || col>board[row].length-1){
            return false;
        }
        if(board[row][col] != word.charAt(index) || emptyBoard[row][col] == 1){
            return false;
        }
        emptyBoard[row][col]++;
        return wordExits(board, emptyBoard,word ,row-1, col, index+1)|| wordExits(board, emptyBoard,word ,row+1, col, index+1)||
                wordExits(board, emptyBoard,word ,row, col-1, index+1)||wordExits(board, emptyBoard,word ,row, col+1, index+1);

 */

File: Python/Leetcode_Python/Backtracking/WordSearch.py
# Time Complexity: O(nm * n^4) || Space Complexity: O(n*m)

from typing import List, Set


def exist(board: List[List[str]], word: str) -> bool:
    if not board:
        return False

    m = len(board)
    n = len(board[0])
    for i in range(m):
        for j in range(n):
            if board[i][j] == word[0] and checker(board, word, 0, i, j, set()):
                return True

    return False


def checker(
    board: List[List[str]], word: str, char: int, r: int, c: int, visited: Set
) -> bool:
    if char >= len(word):
        return True
    if r < 0 or c < 0 or r >= len(board) or c >= len(board[0]):
        return False
    if word[char] != board[r][c] or (r, c) in visited:
        return False
    visited.add((r, c))

    if (
        not checker(board, word, char + 1, r - 1, c, visited)
        and not checker(board, word, char + 1, r + 1, c, visited)
        and not checker(board, word, char + 1, r, c - 1, visited)
        and not checker(board, word, char + 1, r, c + 1, visited)
    ):
        visited.discard((r, c))
        return False
    return True


print(
    exist(
        [["A", "B", "C", "E"], ["S", "F", "E", "S"], ["A", "D", "E", "E"]], "ABCESEEEFS"
    )
)
